neural_modelling/src/neuron/implementations/neuron_impl_stoc_sigma.h

```diff
@@ -0,0 +1,271 @@
+    // Allocate DTCM for neuron array
+    if (sizeof(neuron_impl_t) != 0) {
+        neuron_array = spin1_malloc(n_neurons * sizeof(neuron_impl_t));
+        if (neuron_array == NULL) {
+            log_error("Unable to allocate neuron array - Out of DTCM");
+            return false;
+        }
+    }
+
+    return true;
+}
+
+static inline uint32_t stoc_sigma_ceil_accum(UREAL value) {
+	uint32_t bits = bitsuk(value);
+	uint32_t integer = bits >> 16;
+	uint32_t fraction = bits & 0xFFFF;
+	if (fraction > 0) {
+	    return integer + 1;
+	}
+	return integer;
+}
+
+static inline void neuron_model_initialise(
+		neuron_impl_t *state, neuron_params_t *params) {
+	UREAL ts = params->time_step;
+	state->alpha = params->alpha;
+	state->bias = params->bias;
+	state->t_refract = stoc_sigma_ceil_accum(ukdivuk(params->tau_refract, ts));
+    state->refract_timer = params->refract_init;
+    spin1_memcpy(state->random_seed, params->random_seed, sizeof(mars_kiss64_seed_t));
+    validate_mars_kiss64_seed(state->random_seed);
+    log_info("%u %u %u %u", state->random_seed[0], state->random_seed[1],
+    		state->random_seed[2], state->random_seed[3]);
+    state->inputs[0] = ZERO;
+    state->inputs[1] = ZERO;
+}
+
+static inline void neuron_model_save_state(neuron_impl_t *state, neuron_params_t *params) {
+	params->refract_init = state->refract_timer;
+	spin1_memcpy(params->random_seed, state->random_seed, sizeof(mars_kiss64_seed_t));
+}
+
+SOMETIMES_UNUSED // Marked unused as only used sometimes
+static void neuron_impl_load_neuron_parameters(
+        address_t address, uint32_t next, uint32_t n_neurons,
+        address_t save_initial_state) {
+
+    neuron_params_t *params = (neuron_params_t *) &address[next];
+    for (uint32_t i = 0; i < n_neurons; i++) {
+        neuron_model_initialise(&neuron_array[i], &params[i]);
+    }
+
+    // If we are to save the initial state, copy the whole of the parameters
+    // to the initial state
+    if (save_initial_state) {
+        spin1_memcpy(save_initial_state, address,
+        		n_neurons * sizeof(neuron_params_t));
+    }
+}
+
+SOMETIMES_UNUSED // Marked unused as only used sometimes
+static void neuron_impl_store_neuron_parameters(
+        address_t address, uint32_t next, uint32_t n_neurons) {
+    neuron_params_t *params = (neuron_params_t *) &address[next];
+	for (uint32_t i = 0; i < n_neurons; i++) {
+		neuron_model_save_state(&neuron_array[i], &params[i]);
+	}
+}
+
+SOMETIMES_UNUSED // Marked unused as only used sometimes
+static void neuron_impl_add_inputs(
+        index_t synapse_type_index, index_t neuron_index,
+        input_t weights_this_timestep) {
+    // Get the neuron itself
+    neuron_impl_t *neuron = &neuron_array[neuron_index];
+
+    // Do something to store the inputs for the next state update
+    neuron->inputs[synapse_type_index] += weights_this_timestep;
+}
+
+SOMETIMES_UNUSED // Marked unused as only used sometimes
+static void neuron_impl_do_timestep_update(
```

Comment:
again indexes reused only sdram for 1 recording  and two chances to spike

Reply:
See above